### PR TITLE
fix(android/engine): Fix resetContext calls when selection changes

### LINKED ISSUE #8556: bug(android): "Undo" suggestion is not working Keyman 17.0.80 alpha build
### Describe the bug

I noticed that after pressing the backspace key, the broken suggestion did not appear on the suggestion banner. 

### Reproduce the bug

1. Install Keyman 17.0.80-alpha build. 
2. Open Keyman In-App.
3. Type P r o b
4. Select Problem from the banner. 
5. Press Backspace 

Here, I noticed that the broken suggestion 'Prob' was missing in the suggestion banner. Seems to be an issue. 

I have attached the Screenshot as well as the Video file for reference. 

..broken suggestion Prob is missing

   

..broken suggestion is missing

https://user-images.githubusercontent.com/19683143/229106512-d5ad20d2-1b36-43f7-a2ed-b07de64e3ea9.mp4

### Expected behavior

The broken suggestion 'Prob' should appear on the suggestion banner. 

### Related issues

https://github.com/keymanapp/keyman/issues/7167

### Keyman apps

- [X] Keyman for Android
- [ ] Keyman for iPhone and iPad
- [ ] Keyman for Linux
- [ ] Keyman for macOS
- [ ] Keyman for Windows
- [ ] Keyman Developer
- [ ] KeymanWeb
- [ ] Other - give details at bottom of form

### Keyman version

17.0.80-alpha

### Operating system

Android 11.0, 12.0

### Device

Redmi Note 8 Pro, API 31 emulator

### Target application

_No response_

### Browser

_No response_

### Keyboard name

_No response_

### Keyboard version

_No response_

### Language name

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by darcywong00**
Fixes #8556 and continuing refactor in #7881

> The bug was introduced during the consolidation in #8438 where `KMManager.SystemKeyboardShouldIgnoreSelectionChange` gets set, but never `KMManager.InAppKeyboardShouldIgnoreSelectionChange`. This was causing resetContext to trigger on every change.

I also note these are never set to true, but am not addressing on this PR:
* KMManager.InAppKeyboardShouldIgnoreTextChange
* KMManager.SystemKeyboardShouldIgnoreTextChange

## User Testing
*Setup* - Install the PR build of Keyman for Android and also install the Keyman keyboard kbd_basicfr (Basic French) on an Android device/emulator. Set Keyman as the default system keyboard. Note - if pressing the shift layer deletes selected keys, that's already a known issue #7866.

**GROUP_INAPP** - Verifies inapp keyboard. Use Keyman for Android app
**GROUP_SYSTEM** - Verifies system keyboard. Use Contacts app to add a new contact name (not Chrome/Browser because the searchbar interferes with selection)

* **TEST_UNDO_SUGGESTION** - Verifies backspace after a suggestion works
1. With the sil_euro_latin keyboard, type "prob"
2. Select "problem" from the suggestion banner
3. Press <kbd>Backspace</kbd> key 
4. Verify "prob" is now the left suggestion (the context before accepting the suggestion)
5. Select "prob" from the suggestion banner
6. Verify the text changes from "problem" to "prob"

* **TEST_REPLACE_SELECTION_DEADKEYS** - Verifies keyboard replaces selected text and handles deadkeys
1. In Keyman with the sil_euro_latin keyboard, type "the quick brown fox"
2. Select the letters "row" within the word "brown". (keyboard switches to shift layer)
3. Type "H"
4. Verify the text changes from "brown" to "bHn"
5. Select the letters "ick" within the word "quick"
6. Press <kbd>backspace</kbd> key
7. Verify the text changes from "quick" to "qu"
8. Select the letter "o" within the word "fox"
9. Press <kbd>enter</kbd> key
10. If testing INAPP
  a. verify the word changes from "fox" to "f" followed by "x" on the next line
11. If testing SYSTEM
  a. verify the word changes from "fox" to "fx" (Name is limited to one line)
  b. Select the end of the name field
12. Use the Keyman globe key to switch to the kbd_basicfr keyboard
13. Press the green <kbd>^</kbd>
14. Verify nothing is output (it's a deadkey)
15. Press a vowel
16. Verify the vowel with the combined diacritic is output

### CODE REVIEW

**Comment by mcdurdin** [comment]
> I also note these are never set to true, but am not addressing on this PR:

If these are never set to true, how does this PR fix things? :grin:

**Comment by mcdurdin** [comment]
Do you mean 'never set to false'?

File: android/KMEA/app/src/main/java/com/keyman/engine/KMKeyboardJSHandler.java [hunk 1]
```diff
@@ -149,6 +151,9 @@ public void run() {
 
         // Perform left-deletions
         if (deleteLeft > 0) {
+          if (k.keyboardType == KeyboardType.KEYBOARD_TYPE_INAPP) {
+            KMManager.InAppKeyboardShouldIgnoreSelectionChange = true;
+          }
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/keyman/engine/KMKeyboardJSHandler.java:156]: Why is this only for inapp?

File: android/KMEA/app/src/main/java/com/keyman/engine/KMTextView.java [hunk 2]
```diff
@@ -167,7 +167,7 @@ protected void onTextChanged(CharSequence text, int start, int lengthBefore, int
   protected void onSelectionChanged(int selStart, int selEnd) {
     super.onSelectionChanged(selStart, selEnd);
     if (activeView != null && activeView.equals(this)) {
-      if (KMManager.updateSelectionRange(KMManager.KeyboardType.KEYBOARD_TYPE_INAPP, selStart, selEnd)) {
+      if (KMManager.updateSelectionRange(KMManager.KeyboardType.KEYBOARD_TYPE_INAPP, selStart, selEnd) && !KMManager.InAppKeyboardShouldIgnoreSelectionChange) {
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/keyman/engine/KMTextView.java:170]: This looks like new behaviour? Can you explain why this has changed? (Comparing to stable-16.0)

**Review by mcdurdin — APPROVED** [review]
LGTM I think, if those questions come out positive :grin:

**Comment by darcywong00** [comment]
> If these are never set to true, how does this PR fix things? :grin

...TextChange never gets set. This PR updates Inapp...SelectionChange

File: android/KMEA/app/src/main/java/com/keyman/engine/KMKeyboardJSHandler.java (hunk 1 above)
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/keyman/engine/KMKeyboardJSHandler.java:156]: insertText got consolidated to use the SystemKeyboard handler.
The stable-16.0 had this for left-deletions

https://github.com/keymanapp/keyman/blob/869b7900944cbe85ac1d11fa14e5cc3b29556b7d/android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java#L2764-L2773

I suppose I should also set `InAppKeyboardShouldIgnoreTextChange = true;`

**Comment by bharanidharanj** [comment]
# Test Results

### GROUP_INAPP: - Verifies inapp keyboard. Use Keyman for Android app

* **TEST_UNDO_SUGGESTION (PASSED):** Tested with the attached PR build (Keyman 17.0.89-alpha-test-8611) in Android 12 / API 31 emulator and here is my observation: 1. Verified that 'Prob' appears on the suggestion bar after pressing the Backspace key. 2. Selecting "Prob" from the suggestion banner shows "Prob" on the text input screen.

**Comment by bharanidharanj** [comment]
### GROUP_INAPP: - Verifies inapp keyboard. Use Keyman for Android app

* **TEST_REPLACE_SELECTION_DEADKEYS (PASSED):** Tested this per the instructions with the attached PR build (Keyman 17.0.89-alpha-test-8611) in Android 12 / API 31 emulator and here is my observation: 1. Verified that the text changes from "brown" to "bHn". 2. Verified that the text changes from "quick" to "qu" after pressing the backspace key. 3. Verified that the word changes from "fox" to "f" followed by "x" on the next line. 4. Switched to French Basic keyboard. Verified that pressing the green ^ key outputs nothing on the input text screen. Pressing a vowel 'a' shows the combined diacritic on the screen. Seems to be working as expected.

**Comment by bharanidharanj** [comment]
# Test Results

### GROUP_SYSTEM: - Verifies system keyboard. Use Contacts app to add a new contact name (not Chrome/Browser because the searchbar interferes with selection)

* **TEST_UNDO_SUGGESTION (PASSED):** Tested with the attached PR build (Keyman 17.0.89-alpha-test-8611) in Android 12 / API 31 emulator and here is my observation: 1. Verified that 'Prob' appears on the suggestion bar after pressing the Backspace key. 2. Selecting "Prob" from the suggestion banner shows "Prob" on the text input screen.

   

* **TEST_REPLACE_SELECTION_DEADKEYS (PASSED):** Tested this per the instructions with the attached PR build (Keyman 17.0.89-alpha-test-8611) in Android 12 / API 31 emulator and here is my observation: 1. Verified that the text changes from "brown" to "bHn". 2. Verified that the text changes from "quick" to "qu" after pressing the backspace key. 3. Verified that the word changes from "fox" to "f" followed by "x" on the next line. 4. Switched to French Basic keyboard. Verified that pressing the green ^ key outputs nothing on the input text screen. Pressing a vowel 'a' shows the combined diacritic on the screen. Seems to be working as expected.

File: android/KMEA/app/src/main/java/com/keyman/engine/KMTextView.java (hunk 2 above)
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/keyman/engine/KMTextView.java]: ah, looks like this wasn't needed. Reverted

**Comment by darcywong00** [comment]
Made a tweak to KMText.

@keyaman-test-bot retest GROUP_INAPP

**Comment by bharanidharanj** [comment]
# Test Results

### GROUP_INAPP: - Verifies inapp keyboard. Use Keyman for Android app

* **TEST_UNDO_SUGGESTION (PASSED):** Retested with the attached PR build (Keyman 17.0.89-alpha-test-8611) in Android 12 / API 31 emulator and here is my observation: 1. Verified that 'Prob' appears on the suggestion bar after pressing the Backspace key. 2. Selecting "Prob" from the suggestion banner shows "Prob" on the text input screen.

   


* **TEST_REPLACE_SELECTION_DEADKEYS (PASSED):** Retested this per the instructions with the attached PR build (Keyman 17.0.89-alpha-test-8611) in Android 12 / API 31 emulator and here is my observation: 1. Verified that the text changes from "brown" to "bHn". 2. Verified that the text changes from "quick" to "qu" after pressing the backspace key. 3. Verified that the word changes from "fox" to "f" followed by "x" on the next line. 4. Switched to French Basic keyboard. Verified that pressing the green ^ key outputs nothing on the input text screen. Pressing a vowel 'a' shows the combined diacritic on the screen. Seems to be working as expected.

**Pull request merged**